### PR TITLE
Fix mail sendmail

### PR DESCRIPTION
**Pull request by themancalledjakob**
## Problem

- *To fix email sending* #228 

## Solution

- [x] *adjust .env to use sendmail / postfix for sending emails*
- [x] report the upstream security issue with sendmail

## PR Status

- [x] Code finished and ready to be reviewed/tested
- [ ] The fix/enhancement were manually tested and worked on all tested machines

## Comment
Tested with yunohost 11.2.2

### CODE REVIEW

**Comment by lapineige** [comment]
!testme

**Comment by lapineige** [comment]
What's your feedback on your instance ?

**Comment by themancalledjakob** [comment]
hey @lapineige, sorry for the late response. I am on vacation and can only rarely sneak in some computer time.

> What's your feedback on your instance ?

I am a single user instance :) I did successfully try it with 2 different accounts and email addresses (one mailbox.org, the other yunohost mail). As there are no other users, I simply installed from scratch multiple times and it worked out of the box. Also, fixing a non-working config like this and reloading it with artisan worked immediately.

So, for me, it just worked. Pixelfed was able to send the email, and I received it properly on my mail provider (mailbox.org).

Reading through the internet and the [manpage](https://www.postfix.org/sendmail.1.html), I get the idea that sendmail (postfix) is actually a good way to solve this.

pixelfed's underlying sendmail command is [ '/usr/sbin/sendmail -bs'](https://github.com/pixelfed/pixelfed/blob/93b2dbab17828e883b6103e61306ef43059852d1/config/mail.php#L115) and therefore doing SMTP for us :)

```
-bs    Stand-alone SMTP server mode. Read SMTP commands  from  standard
              input,  and  write responses to standard output.  In stand-alone
              SMTP server mode, mail relaying and other  access  controls  are
              disabled  by  default.  To  enable  them, run the process as the
              [mail_owner](https://www.postfix.org/postconf.5.html#mail_owner) user.
```

So, technically seems legit to use sendmail. However, the command should possibly better be  '/usr/sbin/sendmail -bs --' (notice the last dashes).
There is a possibly scenario, where a user could create an email-address that starts with '-' and therefore inject commands (under SECURITY mentioned in the manpage). Adding the last two dashes prevent this. As Yunohost has afaik the premise to trust users, this is perhaps less of a critical issue? Though, as registrations are open by default, this could leave someone vulnerable out of the box. But my experience here is minimal. I'll investigate this, and if this is really a thing, perhaps a PR to upstream pixelfed could be an idea.

Alternatively, as this is using smtp... perhaps it may be possible to somehow figure out which commands are sent, and use smtp directly after all?

Am I overthinking this?

**Comment by lapineige** [comment]
> sorry for the late response. I am on vacation and can only rarely sneak in some computer time.

Who said there were any hurry ? :wink:

> I did successfully try it with 2 different accounts and email addresses (one mailbox.org, the other yunohost mail)

Great to know it works with Yunohost mail too :)


Thank you for the detailed answer ! Good to know it's working well :)


> So, technically seems legit to use sendmail. However, the command should possibly better be '/usr/sbin/sendmail -bs --' (notice the last dashes).
There is a possibly scenario, where a user could create an email-address that starts with '-' and therefore inject commands (under SECURITY mentioned in the manpage). Adding the last two dashes prevent this. As Yunohost has afaik the premise to trust users, this is perhaps less of a critical issue?

Do these last dashes changes anything in the behaviour ? If it's just working right, let's use it :)

> Though, as registrations are open by default, this could leave someone vulnerable out of the box.

Yeah, let's avoid that risk :)

**Comment by themancalledjakob** [comment]
nice, that makes sense :)

The additional dashes should not change behaviour.
I'm not sure how to add them in the sendmail command in pixelfed without adjusting the github repository. is there maybe a config we could set?
For testing, I quickly made a fork and adjusted the default setting: https://github.com/themancalledjakob/pixelfed 
This is not ideal, as I will very likely never update this fork in the future.

But, good news: when I use this url in the manifest.toml and do a fresh install, everything still works!

**Comment by lapineige** [comment]
> I'm not sure how to add them in the sendmail command in pixelfed without adjusting the github repository.

I don't understand what you mean. Can you explain it again please ? :sweat_smile:

Oh wait… you mean this has to be patched in Pixelfed itself ? https://github.com/themancalledjakob/pixelfed/commit/4311891fe4dcf694fb1ec9439db51298ce9f23cf#diff-af6ac21d51218ff8f16692c0c8a9bf2fb420d3cc0de12458049dd1ed9c5ec244
We could patch it, but if that's a security issue, we should probably report this upstream.

**Comment by themancalledjakob** [comment]
> Oh wait… you mean this has to be patched in Pixelfed itself ?

exactly :)
I am not sure how quick they will patch it though, so we would need a temporary solution. It would be great if we could define a custom sendmail-command, without having to fork the whole pixelfed repository. otherwise we would need to update the fork constantly.
Sorry, I find this super hard to put into words clearly. :upside_down_face: I'll try again:
basically, we can set all kind of settings by simply providing a custom `.env` file, right? ... sooo.. what I'm looking for, is a similar way to provide a custom sendmail-command and still use the upstream pixelfed repository. afaik there is no setting in the `.env` for this, but [they seem to suggest](https://github.com/pixelfed/pixelfed/blob/93b2dbab17828e883b6103e61306ef43059852d1/config/mail.php#L110C56-L111C71) that their sendmail command is a default command, and it sounds as if it can be adjusted by us... but how :detective:  

>  We could patch it, but if that's a security issue, we should probably report this upstream.

yep, I did send them an email to [here](https://github.com/pixelfed/pixelfed/blob/93b2dbab17828e883b6103e61306ef43059852d1/SECURITY.md) :)

**Comment by lapineige** [comment]
> I am not sure how quick they will patch it though, so we would need a temporary solution.

It's easy to patch a line in a file :slightly_smiling_face:

**Comment by themancalledjakob** [comment]
aaaah, of course. you mean simply doing something like:
```
sed -i -e "s#'/usr/sbin/sendmail -bs'#'/usr/sbin/sendmail -bs --'#g" $install_dir/config/mail.php
```
in the install script?

**Comment by lapineige** [comment]
Indeed :)

**Comment by themancalledjakob** [comment]
is there a convention, where to put this and for which actions to send a script progression?
I would intuitively add it to install, upgrade and I guess restore somewhere like this, or should it be silent?
I usually tend to patch source files as soon as possible after downloading them, so that there are not accidentally different versions of files accessed.

```
...

#=================================================
# DOWNLOAD, CHECK AND UNPACK SOURCE
#=================================================
ynh_script_progression --message="Setting up source files..." --weight=1

# Download, check integrity, uncompress and patch the source from app.src
ynh_setup_source --dest_dir="$install_dir"

chmod -R o-rwx "$install_dir"
chown -R $app:www-data "$install_dir"

#=================================================
# PATCHING SOURCE
#=================================================
ynh_script_progression --message="Patching source files..." --weight=1

# Prevent privilege escalation by injecting commands in an email name
# This described in more detail on the manpage https://www.postfix.org/sendmail.1.html under security
ynh_replace_string --match_string="'/usr/sbin/sendmail -bs'" --replace_string="'/usr/sbin/sendmail -bs -- '" --target_file=$install_dir/config/mail.php

#=================================================
# PHP-FPM CONFIGURATION
#=================================================
ynh_script_progression --message="Configuring PHP-FPM..." --weight=1

ynh_add_fpm_config --usage=$fpm_usage --footprint=$fpm_footprint

ynh_add_nginx_config

...
```

**Comment by lapineige** [comment]
Looks good to me !

**Comment by lapineige** [comment]
!testme

**Review by lapineige — APPROVED** [review]

**Comment by lapineige** [comment]
Once continuous integration is happy, feel free to merge :slightly_smiling_face:

**Comment by lapineige** [comment]
I tried it (with password reset), I have a something went wrong :(
And I can't get the details, I get an error code but I don't remember in which Pixelfed site it's possible to translate it…

File: conf/.env [hunk 1]
```diff
@@ -84,7 +84,7 @@ NODEINFO=true
 WEBFINGER=true
 
 ## Mail Configuration (Post-Installer)
-MAIL_DRIVER=smtp
+MAIL_DRIVER=sendmail
 MAIL_HOST=localhost
 MAIL_PORT=25
 MAIL_USERNAME=null
```
Comment (alexAubin) [inline conf/.env:90]: Eeeh by the way, a new mechanism landed in YunoHost 11.2 a couple weeks ago which now allows app to send proper, DKIM-signed mails assuming you do configure the mail username and password ... cf the `allow_email` property in the system user resource : https://yunohost.org/en/contribute/packaging_apps/resources/appresources#system-user

Dunno if that helps but what I do know is that, wether you're using smtp or sendmail, as long as you are not properly authenticated with a login/passwords, emails wont be DKIM-signed and therefore likely to be marked as spam

Comment (lapineige) [reply conf/.env:90]: Thanks for the reminder.

What do we need to use it ?
The mail password and mail username (username@domain.tld I guess ?) ?

Comment (themancalledjakob) [reply conf/.env:90]: @alexAubin yes, we tried this in another PR #229 
the mail user was created, but somehow there were no mails sent.

does maybe something pop in your eye here?
```
## Mail Configuration (Post-Installer)
MAIL_DRIVER=smtp
MAIL_HOST=127.0.0.1
MAIL_PORT=587
MAIL_USERNAME=__APP__
MAIL_PASSWORD=__MAIL_PWD__
MAIL_ENCRYPTION=null
MAIL_FROM_ADDRESS="__APP__@__DOMAIN__"
MAIL_FROM_NAME="Pixelfed"
```

**Comment by themancalledjakob** [comment]
> I tried it (with password reset), I have a something went wrong :( And I can't get the details, I get an error code but I don't remember in which Pixelfed site it's possible to translate it…

ah damn. so that means this PR isn't ready yet, I guess.

You can decrypt the error message like this on your yunohost server in the shell:
```
$ sudo yunohost app shell pixelfed
$ php artisan tinker
Psy Shell v0.11.18 (PHP 8.2.8 — cli) by Justin Hileman
> decrypt('payload here')
```
haven't tried it yet myself actually, but can't wait to get an error message to try it out :-D

possibly you can also do it in the admin panel, like the dev explains [here](https://github.com/pixelfed/docs/issues/77), but i couldn't find the tab they were talking about

File: conf/.env (hunk 1 above)
Comment (lapineige) [reply conf/.env]: I'll try on a sub-PR.

**Comment by lapineige** [comment]
Strangely, if I force upgrade to this version, the env file is not updated.

**Comment by lapineige** [comment]
`Unable to connect with STARTTLS: stream_socket_enable_crypto(): Peer certificate CN=`social.lapineige.fr' did not match expected CN=`localhost'`

**Comment by lapineige** [comment]
I changed the file manually. This is still happening :

> The thing is it's using the wrong domain, it's using the default domain of my server, not Pixelfed domain.

**Comment by lapineige** [comment]
So.... I restarted everything (the new parameters in the .env were not taken into account apparently).

And it works !!

So now we need to handle upgrades to fix this for existing installs.

**Comment by lapineige** [comment]
Done ! 😃

**Comment by lapineige** [comment]
!testme

**Comment by lapineige** [comment]
Thanks a lot for your work !!

**Comment by FatherMcGruder** [comment]
Will email just work with the next update, or will I have to subsequently adjust the .env and/or other system settings?

**Comment by lapineige** [comment]
The `.env` should be overridden during the upgrade (a backup a is created in the same folder), so it should just work.

**Pull request merged**